Added remote debugging configuration

olegz opened the pull request:
Added remote debugging configuration to LocalDeployer which coudl now be set using 'debugPort' and 'debugSuspend' properties

Resolves #35

olegz committed:
Added remote debugging configuration

Added remote debugging configuration to LocalDeployer which coudl now be set using 'debugPort' and 'debugSuspend' properties

Resolves #35

sabbyanandan commented:
Hi, @aclement. A _very long_ time ago, you had tried to add similar support for it via: spring-cloud/spring-cloud-dataflow#369. When you get a chance, it'd be great to hear your thoughts on this.

olegz commented:
@aclement to make the review simpler here is the doc PR hat explains its usage within data-flow https://github.com/spring-cloud/spring-cloud-dataflow/pull/1909/commits/584d148c220d1addd4d47bf7d845ab7b2d0c72b5

markpollack commented on spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/LocalAppDeployer.java:
```diff
@@ -137,6 +137,11 @@ public String deploy(AppDeploymentRequest request) {
 				ProcessBuilder builder = buildProcessBuilder(request, appInstanceEnv, args, Optional.of(i)).inheritIO();
+
+				if (request.getDeploymentProperties().containsKey(LocalDeployerProperties.DEBUG_PORT)) {
+					builder.command().add(1, this.buildRemoteDebugInstruction(request.getDeploymentProperties()));
```
In the original PR from andy there was this tidbit 
```
command.add("-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address="+(debugPort+i));
```
The code is in a loop to deploy the apps based on the deployment property `count` so the port number needs to be difference for each instance that is deployed.  That is they there is the `+i` off the "basePort".

markpollack commented on spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/LocalAppDeployer.java:
```diff
@@ -197,6 +202,23 @@ public void shutdown() throws Exception {
+		String ds = deploymentProperties.getOrDefault(LocalDeployerProperties.DEBUG_SUSPEND, "y");
+		StringBuilder debugCommandBuilder = new StringBuilder();
+		if (StringUtils.hasText(dp)) {
+			logger.warn("***** [REMOTE DEBUGGING] REMOTE DEBUGGING IS ENABLED ON PORT " + dp + " *****");
```
The `deploymentId` should be passed into this method to give context to these warning messages, otherwise one can't tell what app is having remote debugging enabled. I think we should town down the `ALL CAPS` and `****` as well.

markpollack replied:
Actually, while doing some refactoring, this property should also be available for when deploying Tasks....so it should move up into the `AbstractLocalDeployerSupport` class and `LocalTaskLauncher` can take advantage of it as well.

markpollack replied:
I'll push a variation on this to a new branch.

olegz committed:
Added instanceId to debugPort

Added debugPort + instanceId logic to ensure there are no port conflicts in the event there are multiple instances of the application

olegz commented:
#86 includes it
